File: solutions/java/zebra-puzzle/1/src/main/java/ZebraPuzzle.java
import java.util.List;
import java.util.HashMap;
import java.util.function.IntBinaryOperator;

class ZebraPuzzle {
    class CheckError extends Exception {
    }
    
    List<String> color = List.of("red", "green", "yellow", "blue", "ivory");
    List<String> nation = List.of("Englishman", "Spaniard", "Ukrainian", "Norwegian", "Japanese");
    List<String> pet = List.of("dog", "snail", "fox", "horse", "zebra");
    List<String> beverage = List.of("coffee", "tea", "milk", "juice", "water");
    List<String> hobby = List.of("dancing", "painter", "reading", "football", "chess");
    List<List<String>> chosen = List.of(color, nation, pet, beverage, hobby);
    HashMap<String, Integer> house = new HashMap<String, Integer>();

    ZebraPuzzle() {
        dfs(0, 0);
    }

    void checkOnly(String inHouse, String outHouse) throws CheckError {
        for (List<String> category: chosen)
            if (category.contains(outHouse)) {
                for (String other: category)
                    if (house.containsKey(other) && house.get(inHouse) == house.get(other))
                        throw new CheckError();
                return;
            }
    }

    void check(String x, String y) throws CheckError {
        if (!house.containsKey(x) && !house.containsKey(y))
            return;
        else if (house.containsKey(x) && !house.containsKey(y))
            checkOnly(x, y);
        else if (!house.containsKey(x) && house.containsKey(y))
            checkOnly(y, x);
        else if (!house.get(x).equals(house.get(y)))
            throw new CheckError();
    }

    void check(String x, String y, IntBinaryOperator func) throws CheckError {
        if (!house.containsKey(x) || !house.containsKey(y))
            return;
        if (func.applyAsInt(house.get(x), house.get(y)) != 0)
            throw new CheckError();
    }

    boolean valid() {
        try {
            check("Englishman", "red");
            check("Spaniard", "dog");
            check("green", "coffee");
            check("Ukrainian", "tea");
            check("green", "ivory", (x, y) -> x - (y + 1));
            check("snail", "dancing");
            check("yellow", "painter");
            check("milk", "milk", (x, y) -> x - 2);
            check("Norwegian", "Norwegian", (x, y) -> (x));
            check("reading", "fox", (x, y) -> Math.abs(x - y) - 1);
            check("painter", "horse", (x, y) -> Math.abs(x - y) - 1);
            check("football", "juice");
            check("Japanese", "chess");
            check("Norwegian", "blue", (x, y) -> Math.abs(x - y) - 1);
        }
        catch (CheckError e) {
            return false;
        }
        return true;
    }

    boolean dfs(int depHouse, int depChosen) {
        if (depHouse == 5)
            return true;
        if (depChosen == 5)
            return dfs(depHouse + 1, 0);
        if (!valid())
            return false;
        for (String now: chosen.get(depChosen)) {
            if (house.containsKey(now))
                continue;
            house.put(now, depHouse);
            if (dfs(depHouse, depChosen + 1))
                return true;
            house.remove(now);
        }
        return false;
    }
    
    String getWaterDrinker() {
        for (String key: house.keySet())
            if (house.get("water").equals(house.get(key)) && nation.contains(key))
                return key;
        return null;
    }

    String getZebraOwner() {
        for (String key: house.keySet())
            if (house.get("zebra").equals(house.get(key)) && nation.contains(key))
                return key;
        return null;
    }
}

File: solutions/cpp/zebra-puzzle/1/zebra_puzzle.cpp
#include "zebra_puzzle.h"

#include <array>
#include <algorithm>
#include <cmath>
#include <string>

constexpr int N = 5;

enum { ENGLISH, SPANIARD, UKRAINIAN, NORWEGIAN, JAPANESE };
enum { RED, GREEN, IVORY, YELLOW, BLUE };
enum { DOG, SNAIL, FOX, HORSE, ZEBRA };
enum { COFFEE, TEA, MILK, JUICE, WATER };
enum { READING, DANCING, PAINTING, FOOTBALL, CHESS };

static const std::array<std::string, N> RESIDENTS = {
    "Englishman", "Spaniard", "Ukrainian", "Norwegian", "Japanese"
};

static int owner_of(const std::array<int, N>& resident_to_house, int house)
{
    for (int i = 0; i < N; ++i)
        if (resident_to_house[i] == house)
            return i;
    return -1;
}

namespace zebra_puzzle {

Solution solve()
{
    Solution solution{};

    std::array<int, N> color   = {0,1,2,3,4},
                       resident= {0,1,2,3,4},
                       pet     = {0,1,2,3,4},
                       drink   = {0,1,2,3,4},
                       hobby   = {0,1,2,3,4};

    do {
        // Colors
        if (std::abs(color[GREEN] - color[IVORY]) != 1)
            continue;

        do {
            // Residents
            if (resident[NORWEGIAN] != 0) continue;
            if (std::abs(resident[NORWEGIAN] - color[BLUE]) != 1) continue;
            if (resident[ENGLISH] != color[RED]) continue;

            do {
                // Pets
                if (resident[SPANIARD] != pet[DOG]) continue;

                do {
                    // Drinks
                    if (drink[COFFEE] != color[GREEN]) continue;
                    if (resident[UKRAINIAN] != drink[TEA]) continue;
                    if (drink[MILK] != 2) continue;

                    do {
                        // Hobbies
                        if (pet[SNAIL] != hobby[DANCING]) continue;
                        if (color[YELLOW] != hobby[PAINTING]) continue;
                        if (hobby[FOOTBALL] != drink[JUICE]) continue;
                        if (resident[JAPANESE] != hobby[CHESS]) continue;
                        if (std::abs(hobby[READING] - pet[FOX]) != 1) continue;
                        if (std::abs(hobby[PAINTING] - pet[HORSE]) != 1) continue;

                        // Solution found
                        solution.drinksWater = RESIDENTS[owner_of(resident, drink[WATER])];
                        solution.ownsZebra = RESIDENTS[owner_of(resident, pet[ZEBRA])];

                        return solution;

                    } while (std::next_permutation(hobby.begin(), hobby.end()));

                } while (std::next_permutation(drink.begin(), drink.end()));

            } while (std::next_permutation(pet.begin(), pet.end()));

        } while (std::next_permutation(resident.begin(), resident.end()));

    } while (std::next_permutation(color.begin(), color.end()));

    return solution;
}

}  // namespace zebra_puzzle





